engine/rpcserver.go

```diff
@@ -1213,3 +1221,322 @@ func (s *RPCServer) GetAuditEvent(ctx context.Context, r *gctrpc.GetAuditEventRe
+	if !gctscript.GCTScriptConfig.Enabled {
+		return &gctrpc.GCTScriptGenericResponse{Status: gctscript.ErrScriptingDisabled.Error()}, nil
+	}
+
+	UUID, err := uuid.FromString(r.Script.UUID)
+	if err != nil {
+		return &gctrpc.GCTScriptGenericResponse{Status: "error", Data: err.Error()}, nil
+	}
+
+	if v, f := gctscript.AllVMs[UUID]; f {
+		err = v.Shutdown()
+		status := " terminated"
+		if err != nil {
+			status = err.Error()
+		}
+		return &gctrpc.GCTScriptGenericResponse{Status: "ok", Data: v.ID.String() + status}, nil
+	}
+	return &gctrpc.GCTScriptGenericResponse{Status: "error", Data: "no running script found"}, nil
+}
+
+// GCTScriptUpload upload a new script to ScriptPath
+func (s *RPCServer) GCTScriptUpload(ctx context.Context, r *gctrpc.GCTScriptUploadRequest) (*gctrpc.GCTScriptGenericResponse, error) {
+	if !gctscript.GCTScriptConfig.Enabled {
+		return &gctrpc.GCTScriptGenericResponse{Status: gctscript.ErrScriptingDisabled.Error()}, nil
+	}
+
+	fPath := filepath.Join(gctscript.ScriptPath, r.ScriptName)
+	_, err := os.Stat(fPath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+	} else {
+		if !r.Overwrite {
+			return nil, fmt.Errorf("%s script found and overwrite set to false", r.ScriptName)
+		} else if r.Overwrite {
+			f := filepath.Join(gctscript.ScriptPath, "version_history")
+			err = os.MkdirAll(f, 0770)
+			if err != nil {
+				return nil, err
+			}
+			timeString := strconv.FormatInt(time.Now().UnixNano(), 10)
+			renamedFile := filepath.Join(f, timeString+"-"+r.ScriptName)
+			err = file.Move(fPath, renamedFile)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	newFile, err := os.Create(fPath)
+	if err != nil {
+		return nil, err
+	}
+	defer newFile.Close()
+
+	_, err = newFile.Write(r.Data)
+	if err != nil {
+		return nil, err
+	}
+	err = newFile.Close()
+	if err != nil{
+		log.Errorln(log.Global, "failed to close file handle, archive removal may fail")
+	}
+
+	if r.Archived {
+		z, errZip := zip.OpenReader(fPath)
+		if errZip != nil {
+			return nil, err
+		}
+
+		defer z.Close()
+
+		for x := range z.Reader.File {
+			var zFile io.ReadCloser
+			zFile, err = z.Reader.File[x].Open()
+			if err != nil {
+				return nil, err
+			}
+
+			zipPath := filepath.Join(gctscript.ScriptPath, z.Reader.File[x].Name) // nolint:gosec
+			// We ignore gosec linter above because the code below files the file traversal bug when extracting archives
+			if !strings.HasPrefix(zipPath, filepath.Clean(gctscript.ScriptPath)+string(os.PathSeparator)) {
+				_ = zFile.Close()
+				return nil, fmt.Errorf("%s: illegal file path", fPath)
+			}
+
+			if z.Reader.File[x].FileInfo().IsDir() {
+				err = os.MkdirAll(zipPath, z.Reader.File[x].Mode())
+				if err != nil {
+					return nil, err
+				}
+			} else {
+				var outFile *os.File
+				outFile, err = os.OpenFile(zipPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, z.Reader.File[x].Mode())
+				if err != nil {
+					return nil, err
+				}
+
```

Comment:
We Validate an individual script but it would be nice if we could also Validate each individual file inside the archive and report of an issue before we write the contents to the gctscript dir

Reply:
I had this discussion with @shazbert awhile back could not ever think of a clean solution 

scripts have the ability to import other scripts (for example you might have a  helper function imported by multiple scripts) I figured these would be the type of things to be commonly uploaded in archived format and i wasn't sure the best way to handle telling the validator which is the "main" script

For example i have 
Hello.zip
-> main.gct
-> helper_funcs.gct

if it tries to execute helper_funcs.gct it might not build because it has nothing that can be executed without being called from main.gct i would need some way to tell it what is the entry point 

I thought of:

1) file inside zip that is parsed like entry.txt that will just have the main file name in it to execute

2) a -skipvalidation flag that you can pass if you ever run into this issue (this might be a good idea to have either way)


Open to all suggestions

Reply:
There should not ever be a case where  helper fails unless it executes code that it needs from another script

a script may fail if it imports a helper that has not been written yet but with the new zip file changes i just pushed i am making a change now to extract all files and execute each one in the validator should have that pushed up shortly

Reply:
the upload support has been reworked
it will now validate all files extracted by the zip file and report any failures

Reply:
while this feature is almost complete after a discussion yesterday I have decided to make one more change and that is to archive folders before moving them to the version_history because while running in docker there is a chance we might run into the os.Rename bug if people are trying to move cross partition and currently the file.Move package does not support moving entire folders currently due to it using io.Copy and performance issues with this

Reply:
With the merge of #411 this change has been completed
if there is a folder with the same name as a archive you are trying to upload it will archive it will first zip and move to the version_history folder